### ISSUE TITLE
Monitoring API corrupts output when id contains environment variable reference

### ISSUE THREAD

**Issue by yaauie** [elastic/logstash#10546]
With the following pipeline:

~~~
input {
  stdin {
    id => "${LS_INPUT_ID}"
  }
}
output {
  stdout { }
}
~~~

Invoked with:

~~~
 LS_INPUT_ID=foo bin/logstash -f envvar.conf
~~~

The stats API shows two inputs, each of which has only a portion of the relevant data, one with and one without environment variable expansion:
~~~
curl --silent http://localhost:9600/_node/stats/pipelines | jq .pipelines.main.plugins.inputs
[
  {
    "id": "foo",
    "events": {
      "out": 0,
      "queue_push_duration_in_millis": 0
    }
  },
  {
    "id": "${LS_INPUT_ID}",
    "name": "stdin"
  }
]
~~~

When we create the same pipeline with a literal id, we get a single input that contains `id`, `name`, and `events`.
~~~
curl --silent http://localhost:9600/_node/stats/pipelines | jq .pipelines.main.plugins.inputs
[
  {
    "id": "foo",
    "name": "stdin",
    "events": {
      "out": 0,
      "queue_push_duration_in_millis": 0
    }
  }
]
~~~

**Comment by yaauie** [comment]
I _suspect_ that we are using the LIR Graph's `Vertex#getId()` (which is _not_ expanded) to set the metadata, and `Plugin#id` (which _is_ expanded) to set the ongoing stats.

**Comment by jsvd** [comment]
and for outputs this field isn't even replaced:

`TEST=hello bin/logstash -e 'input { stdin { id => "${TEST}_in" } } output { stdout {id => "${TEST}_out" } }'`

results in:

```
% curl -s localhost:9600/_node/stats | jq ".pipelines.main.plugins.inputs"
[
  {
    "id": "hello_in",
    "events": {
      "out": 0,
      "queue_push_duration_in_millis": 0
    }
  },
  {
    "id": "${TEST}_in",
    "name": "stdin"
  }
]
% curl -s localhost:9600/_node/stats | jq ".pipelines.main.plugins.outputs"
[
  {
    "id": "${TEST}_out",
    "name": "stdout",
    "events": {
      "in": 0,
      "out": 0,
      "duration_in_millis": 235
    }
  }
]
```

**Comment by andsel** [comment]
I think this problem rises in the creation of codec done in these 3 points:
- for inputs https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java#L190-L191
- for filters https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java#L174-L175
- for outputs https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java#L157-L157

`convertArgs` create the codec for Ruby plugins and `convertJavaArgs` does the same for Java plugins. The difference in the 2 methods is that the Java part seems to expand the variable using `ConfigVariableExpander` while the Ruby path not. To prove it we could test a Java pipeline with codec with a placeholder to be substituted. One time we use a Java codec and the other a Ruby codec. If for the Java codec the `CompiledPipeline` replace it and for the other not, we should have the proof that the bug reside there.

**Comment by andsel** [comment]
My guess was wrong, else any Java pipeline with Ruby plugins could ever worked with env vars substitution. The plugin id is brought from the LIR graph, https://github.com/elastic/logstash/blob/5f50b7f18d1323bed69b272da9d37cdef2514637/logstash-core/src/main/java/org/logstash/plugins/PluginFactoryExt.java#L186-L190 and at that step (before plugin factory instantiate the plugins) no replacement happens, at least it seems.

**Comment by andsel** [comment]
Closed by PR #11592
